pkg/skaffold/docker/default_repo.go

```diff
@@ -17,41 +17,40 @@ limitations under the License.
 package docker
 
 import (
-	"regexp"
 	"strings"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker/registry"
 )
 
 const maxLength = 255
 
-var (
-	escapeRegex = regexp.MustCompile(`[/._:@]`)
-	prefixRegex = regexp.MustCompile(`(.*\.)?gcr.io/[a-zA-Z0-9-_]+/?`)
-)
-
-func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, error) {
+func SubstituteDefaultRepoIntoImage(defaultRepo string, image string, rewriteStrategy bool) (string, error) {
 	if defaultRepo == "" {
 		return image, nil
 	}
-
 	parsed, err := ParseReference(image)
 	if err != nil {
 		return "", err
 	}
 
-	replaced := replace(defaultRepo, parsed.BaseName)
+	var replaced string
+	if rewriteStrategy {
+		replaced = substituteDefaultRepoIntoImageWithRewrite(defaultRepo, parsed.BaseName)
+	} else {
+		replaced = replace(defaultRepo, parsed.BaseName)
+	}
 	if parsed.Tag != "" {
 		replaced = replaced + ":" + parsed.Tag
 	}
 	if parsed.Digest != "" {
 		replaced = replaced + "@" + parsed.Digest
 	}
-
-	return replaced, nil
+	return truncate(replaced), nil
```

Comment:
The truncation should be on the image name, not the full ref.  That is, the tag + digest don't count.  

Hmm... re-reading the [OCI distribution spec](https://github.com/opencontainers/distribution-spec/blob/master/spec.md#overview) it's actually the repository name portion that must be less than 256 — the registry hostname doesn't count! 🙀